hooks/pre/31-mongo_storage_engine.rb

```diff
@@ -0,0 +1,53 @@
+pulp_db_param = param('katello', 'pulp_db_name') || param('foreman_proxy_content', 'pulp_db_name')
+fail_and_exit('Unknown pulp database name') unless pulp_db_param
+MONGO_DB_NAME = pulp_db_param.value
+
+def fail_and_exit(message)
+  Kafo::Helpers.log_and_say :error, message
+  kafo.class.exit 1
+end
+
+def migration
+  export_dir = '/var/tmp/engine_upgrade'
+  mongo_dir = '/var/lib/mongodb'
+  hiera_file = '/etc/foreman-installer/custom-hiera.yaml'
+  mongo_backup = '/var/tmp/mongo_backup'
+
+  # Create export directory and dump MongoDB
+  logger.info 'Stopping Pulp services except MongoDB'
+  Kafo::Helpers.execute('katello-service stop --exclude rh-mongodb34-mongod')
+  Dir.mkdir_p(export_dir)
+  File.chmod(0700, export_dir)
+  logger.info "Starting mongodump to #{export_dir}"
+  Kafo::Helpers.execute("mongodump --host localhost --out #{export_dir}")
+
+  # Move datafiles out of MongoDB directory so it will start
+  logger.info 'Export done, stopping MongoDB to remove old datafiles'
+  Kafo::Helpers.execute('systemctl stop rh-mongodb34-mongod')
+  logger.info "Moving contents from #{mongo_dir} to #{mongo_backup}"
+  Dir.mkdir_p(mongo_backup)
+  File.chmod(0700, mongo_backup)
+  Kafo::Helpers.execute("mv #{mongo_dir}/* #{mongo_backup}")
+
+  # Start MongoDB and import the dump, fail and notify user of backup if restore does not work.
+  logger.info 'Turning on MongoDB with wiredTiger and starting import'
+  Kafo::Helpers.execute('katello-service start --only rh-mongodb34-mongod')
+  Kafo::Helpers.execute("mongod --storageEngine wiredTiger --dbpath #{mongo_dir}")
+  Kafo::Helpers.execute("mongorestore --host localhost --db #{MONGO_DB_NAME} --quiet --dir #{export_dir}")
+  unless $?.success?
+    fail_and_exit "The restore could not be completed correctly, the backup is located at #{export_dir}"
+  end
+
+  # Remove old data files
+  logger.info "Import done, removing old data in #{mongo_backup}"
+  Kafo::Helpers.execute("rm -rf #{mongo_backup}")
+
+  # Update Hiera to wiredTiger for installer run
+  logger.info 'Changing custom Hiera to use wiredTiger for installer Puppet run.'
+  Kafo::Helpers.execute("sed -i -e 's/run the installer with --upgrade-mongo-storage to upgrade to wiredTiger./Do not remove'/g #{hiera_file}")
+  Kafo::Helpers.execute("sed -i -e 's/mmapv1/wiredTiger/g' #{hiera_file}")
```

Comment:
By having it in we can enforce the installer to keep the engine param in wiredTiger otherwise if a user goes in and changes it to mmapv1 it wont start correctly.

Reply:
As we talked about on IRC, if we are keeping the mongo config file in a controlled state with Puppet, we should keep the db storage engine variable controlled as well.

Reply:
I'm not sure I follow. Puppet does control it. Are you saying that you don't want to rely on the built in default?

Reply:
The built in default is not setting a value, so if we are managing most of the config shouldn't we manage the storage engine too going forward so that if a user changes anything the installer will revert it back to the correct value unless they override it again with hiera?

Reply:
I guess it's ok to put an explicit value in the config. I just don't like making changes to custom-hiera.yaml. That should be a user thing.

Reply:
I'd rather configure this similarly to how we update values in migrations, it might require a bit more effort but I too don't really like running `sed` on a users config file.

Reply:
When we initially starting doing this we looked at storing it in answers but that'll expose the mongodb module at the top level and that's not something we want. In the hook context we have `get_custom_config` and `store_custom_config` (which is cleaner than the marker files but we haven't tried); in this case it won't work because AFAIK that's not passed to the module in hiera.